src/circuit.rs

```diff
@@ -41,6 +42,69 @@ impl From<JubJubAffine> for PublicInputValue {
     }
 }
 
+#[derive(Debug, Clone)]
+/// Collection of structs/objects that the Verifier will use in order to
+/// de/serialize data needed for Circuit proof verification.
+/// This structure can be seen as a link between the [`Circuit`] public input
+/// positions and the [`VerifierKey`] that the Verifier needs to use.
+pub struct VerifierData {
+    pub(self) key: VerifierKey,
+    pub(self) pi_pos: Vec<usize>,
+}
+
+impl VerifierData {
+    /// Creates a new [`VerifierData`] from a [`VerifierKey`] and the public
+    /// input positions of the circuit that it represents.
+    pub const fn new(key: VerifierKey, pi_pos: Vec<usize>) -> Self {
+        Self { key, pi_pos }
+    }
+
+    /// Returns a reference to the contained [`VerifierKey`].
+    pub const fn key(&self) -> &VerifierKey {
+        &self.key
+    }
+
+    /// Returns a reference to the contained Public Input positions.
+    pub const fn pi_pos(&self) -> &Vec<usize> {
+        &self.pi_pos
+    }
+
+    /// Deserializes the [`VerifierData`] into a vector of bytes.
+    #[allow(unused_must_use)]
+    pub fn to_var_bytes(&self) -> Vec<u8> {
+        let mut buff =
+            vec![
+                0u8;
+                VerifierKey::SIZE + u32::SIZE + self.pi_pos.len() * u32::SIZE
+            ];
+        let mut writer = &mut buff[..];
+
+        writer.write(&self.key.to_bytes());
+        writer.write(&(self.pi_pos.len() as u32).to_bytes());
+        self.pi_pos.iter().copied().for_each(|pos| {
+            // Omit the result since disk_bytes write can't fail here
+            // due to the fact that we're writing into a vector basically.
+            let _ = writer.write(&(pos as u32).to_bytes());
+        });
+
+        buff
+    }
+
+    /// Serializes [`VerifierData`] from a slice of bytes.
+    pub fn from_slice(buf: &[u8]) -> Result<Self, Error> {
+        let mut buffer = &buf[..];
+        let key = VerifierKey::from_reader(&mut buffer)?;
+        let pos_num = u32::from_reader(&mut buffer)? as usize;
```

Comment:
No circuits bigger than 1<<32 will appear. At least for some time. We can always upgrade when we need it. But I think it makes no sense to consume twice the space when our biggest circuit is 1<<18. Which is considerably smaller..

Reply:
We are imposing an assumption (circuits < 2^32) on the premise of saving 4 bytes. I don't think that is a good trade-off

Reply:
Well, we could do `u128` if we follow your thought. Being realistic, I've not seen any circuit bigger than 1<<18. And that's indeed really big. We've room to do `14x` that. So again, it;s not only 4 bytes. Is whether is logical to support bigger stuff.

Reply:
It doesn't apply to `u128` because the PI pos is a usize and is `u64` in conventional platforms. So its never gonna hold a `u128`

Reply:
If there are no particular reason to do so, I would prefer use `u32`: remember that currently the `usize` in WASM is `u32` (since it's a 32bit platform), so the vector of public input positions we're using in WebAssembly is u32.

Of course if we have use case currently where we have u64 we can rethink that, but otherwise I would keep as is.